lib/puma/request.rb

```diff
@@ -103,16 +103,29 @@ def handle_request(client, lines, requests)
         res_info[:content_length] = nil
         res_info[:no_body] = head
 
-        res_info[:content_length] = if res_body.kind_of? Array and res_body.size == 1
-          res_body[0].bytesize
-        else
-          nil
+        str_headers(env, status, headers, res_info, lines, requests, client)
+
+        if !res_info[:content_length] && !res_info[:transfer_encoding] && status != 204
+          if res_body.respond_to?(:to_ary)
+            if array_body = res_body.to_ary
+              res_info[:content_length] = array_body.map(&:bytesize).inject(0, :+)
+            elsif res_body.respond_to?(:each)
```

Comment:
This `to_ary` exists but returns `nil` is for handling objects like Rails' RackBody